buffer/src/main/java/io/netty/buffer/ReadOnlyByteBufferBuf.java

```diff
@@ -564,12 +564,80 @@ private ByteBuffer nioBuffer(boolean internal) {
 
     @Override
     public ByteBuf duplicate() {
-        return new ReadOnlyByteBufferBuf(allocator, buffer);
+        return new ReadOnlyDuplicatedByteBuf(this);
     }
 
     @Override
     public ByteBuf slice(int index, int length) {
-        return new ReadOnlyByteBufferBuf(allocator,
-                (ByteBuffer) buffer.duplicate().position(index).limit(index + length));
+        return new ReadOnlySlicedByteBuf(this, index, length);
+    }
+
+    @Override
+    public ByteBuf asReadOnly() {
+        return this;
+    }
+
+    @SuppressWarnings("deprecation")
+    private static final class ReadOnlySlicedByteBuf extends SlicedByteBuf {
+        ReadOnlySlicedByteBuf(ByteBuf buffer, int index, int length) {
+            super(buffer, index, length);
+        }
+
+        @Override
+        public ByteBuf asReadOnly() {
+            return this;
+        }
+
+        @Override
+        public ByteBuf slice(int index, int length) {
+            return new ReadOnlySlicedByteBuf(this, index, length);
+        }
+
+        @Override
+        public ByteBuf duplicate() {
+            return new ReadOnlyDuplicatedByteBuf(this);
+        }
+
+        @Override
+        public boolean isWritable() {
+            return false;
+        }
+
+        @Override
+        public boolean isWritable(int numBytes) {
+            return false;
+        }
+    }
```

Comment:
In addition to the above methods, nioBuffer and nioBuffers methods should return read-only ByteBuffer instances. While looking into this I noticed that the override for nioBuffer exists in ReadOnlyByteBufferBuf, but the nioBuffers override is missing.

Something like this for making nioBuffer and nioBuffers return read-only ByteBuffers?
```java
    @Override
    public ByteBuffer nioBuffer(int index, int length) {
        return unwrap().nioBuffer(index, length).asReadOnlyBuffer();
    }

    @Override
    public ByteBuffer[] nioBuffers(int index, int length) {
        ByteBuffer[] byteBuffers = unwrap().nioBuffers(index, length);
        for (int i = 0; i < byteBuffers.length; i++) {
            byteBuffers[i] = byteBuffers[i].asReadOnlyBuffer();
        }
        return byteBuffers;
    }
```

Reply:
The reason why read-only ByteBuffers are needed in the SslHandler use case is to avoid mutating the buffer. JDK's SSLEngine will mutate the input buffer with an in-place encryption unless the buffer is read-only. The support for read-only ByteBuffers in JDK's SSLEngine was added as a result of a bug reported by @chrisvest, more details in comment https://github.com/netty/netty/issues/14069#issuecomment-2129279047 .